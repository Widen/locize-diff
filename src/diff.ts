import { getInput } from '@actions/core'
import { ResourceCollection } from './types'

interface ResourceDiff {
  key: string
  diffs: Record<string, { left: string; right: string }>
}

function getKeys(left: ResourceCollection, right: ResourceCollection) {
  const keys = Object.keys(right.resources)

  // When including deleted keys, we need to also include the keys from the left
  // side of the comparison. Otherwise, we only include keys from the right side.
  if (getInput('ignoreDeletedKeys') !== 'true') {
    keys.push(...Object.keys(left.resources))
  }

  return keys
}

function diffCollection(
  left: ResourceCollection,
  right: ResourceCollection
): ResourceDiff {
  const diffs = getKeys(left, right)
    .filter((key) => left.resources[key] !== right.resources[key])
    .reduce((acc, key) => {
      acc[key] = {
        left: left.resources[key],
        right: right.resources[key],
      }

      return acc
    }, {} as ResourceDiff['diffs'])

  return { key: left.key, diffs }
}

export function diffResources(
  left: ResourceCollection[],
  right: ResourceCollection[]
): ResourceDiff[] {
  return left.map((leftItem) =>
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    diffCollection(leftItem, right.find((item) => leftItem.key === item.key)!)
  )
}
