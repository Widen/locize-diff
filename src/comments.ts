import { ResourceDiff } from './diff'

export function createMessage(diffs: ResourceDiff[]): string {
  return 'Here are the diffs'

  // ## `en-US/translation`

  // | Key | `latest` | `production` |
  // | --- | --- | --- |
  // | accept | Accept | Accepted |
  // | product-history | Product History | Product history |
  // | product-history | Product History | _No value_ |
}
