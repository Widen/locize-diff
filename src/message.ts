import { getInput } from '@actions/core'
import { ResourceDiff } from './diff'

export const credits =
  '_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._'

function createDiffRow([key, value]: [string, ResourceDiff['diffs'][number]]) {
  return `| \`${key}\` | ${value.left || ''} | ${value.right || ''} |`
}

function createDiffMessage(diff: ResourceDiff) {
  const leftVersion = getInput('leftVersion')
  const rightVersion = getInput('rightVersion')

  const rows = Object.entries(diff.diffs)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(createDiffRow)
    .join('\n')

  return `### \`${diff.key}\`

| Key | \`${leftVersion}\` | \`${rightVersion}\` |
| --- | --- | --- |
${rows}`
}

export function createMessage(diffs: ResourceDiff[]) {
  const projectSlug = getInput('projectSlug')
  const link = projectSlug
    ? ` [Click here](https://locize.app/p/${projectSlug}/settings) to go to your project settings to publish your changes.`
    : ''

  return `Heads up! Looks like there are some differences between your two Locize versions.${link}

${diffs.map(createDiffMessage).join('\n\n')}

${credits}`
}
