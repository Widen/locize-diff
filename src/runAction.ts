import { setFailed } from '@actions/core'
import { context } from '@actions/github'
import { runCommand } from './commands'
import { runDiff } from './commands/diff'

export async function runAction() {
  try {
    if (context.eventName === 'issue_comment') {
      await runCommand()
    } else {
      await runDiff()
    }
  } catch (err) {
    setFailed(err.message)
  }
}
