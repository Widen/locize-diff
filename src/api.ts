import { getInput } from '@actions/core'
import { HttpClient } from '@actions/http-client'
import { BearerCredentialHandler } from '@actions/http-client/auth'
import { Dict, ResourceCollection } from './utils/types'

const authHandler = new BearerCredentialHandler(getInput('apiKey'))
const client = new HttpClient(undefined, [authHandler])

export function getJSON<T>(url: string): Promise<T> {
  return client
    .get(url)
    .then((res) => res.readBody())
    .then((res) => JSON.parse(res))
}

interface ListResourcesResponse {
  key: string
  lastModified: string
  size: number
  url: string
}

export async function listResources(
  projectId: string,
  version: string
): Promise<ListResourcesResponse[]> {
  return getJSON(`https://api.locize.app/download/${projectId}/${version}`)
}

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)
    const url = resource.url.replace(projectId, `pull/${projectId}`)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(url),
    }
  })

  return Promise.all(promises)
}

export async function updateTranslations(
  key: string,
  updates: Record<string, string | null>
) {
  const projectId = getInput('projectId')
  const version = getInput('rightVersion')

  console.log(
    'Updating',
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    updates
  )

  await client.post(
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    JSON.stringify(updates)
  )
}
