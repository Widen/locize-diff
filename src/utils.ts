import { getJSON, listResources } from './api'
import { Dict, ResourceCollection } from './types'

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(resource.url),
    }
  })

  return Promise.all(promises)
}
