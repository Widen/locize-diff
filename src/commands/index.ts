import { context } from '@actions/github'
import { createComment } from '../utils/comments'
import { runDiff } from './check'
import { runCopy } from './copy'

async function reportResult(message?: string) {
  if (message) {
    await createComment(`@${context.actor} ${message}`)
  }
}

export async function runCommand() {
  const comment = context.payload.comment?.body ?? ''
  const match = comment.match(/@locize-diff\s(check|copy)/)
  const command = match?.[1]

  switch (command) {
    case 'check':
      return reportResult(await runDiff())

    case 'copy':
      return reportResult(await runCopy())
  }
}
