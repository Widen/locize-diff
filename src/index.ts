import { getInput, setFailed } from '@actions/core'
import { context, getOctokit } from '@actions/github'
import { getComment, minimizeComment } from './comments'
import { diffResources } from './diff'
import { createMessage } from './message'
import { collectResources } from './utils'

async function main() {
  const projectId = getInput('projectId')
  const leftVersion = getInput('leftVersion')
  const rightVersion = getInput('rightVersion')
  const includeDrafts = getInput('includeDrafts') !== 'true'

  if (includeDrafts && context.payload.pull_request?.draft) {
    return
  }

  try {
    const octokit = getOctokit(getInput('token'))
    const left = await collectResources(projectId, leftVersion)
    const right = await collectResources(projectId, rightVersion)
    const diffs = diffResources(left, right)
    const comment = await getComment()

    if (diffs.length) {
      const req = {
        ...context.repo,
        body: createMessage(diffs),
        issue_number: context.issue.number,
      }

      if (comment) {
        if (comment.body !== req.body) {
          await octokit.issues.updateComment({ ...req, comment_id: comment.id })
        }
      } else {
        await octokit.issues.createComment(req)
      }
    }

    // If the comment exists and there are no longer any diffs, we minimize the
    // comment so it no longer shows in the GitHub UI.
    if (comment && !diffs.length) {
      await minimizeComment(comment.node_id)
    }
  } catch (err) {
    setFailed(err.message)
  }
}

main()
