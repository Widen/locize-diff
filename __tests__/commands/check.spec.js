jest.mock('@actions/core')
jest.mock('@actions/github')
jest.mock('@actions/http-client')

import {
  contextMock,
  createCommentMock,
  graphqlMock,
  listCommentsMock,
  prComment,
  updateCommentMock,
} from '@actions/github'
import { runAction } from '../../src/runAction'
import { mockFetchResource, mockListResources, sampleComment } from '../utils'

beforeEach(() => {
  jest.resetAllMocks()
  contextMock.mockReturnValue(prComment('@locize-diff check'))
})

it('should allow manually requesting a diff', async () => {
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'baz', hi: 'ho' } }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})

it('should notify the user when updating an existing comment', async () => {
  listCommentsMock.mockReturnValue([{ body: sampleComment }])
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar', hi: 'ho' } },
    { 'en-US/translation': { foo: 'baz', hi: 'howdy' } }
  )

  await runAction()

  // Updates the diff with the new diff
  expect(updateCommentMock).toHaveBeenCalledTimes(1)
  expect(updateCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(updateCommentMock.mock.calls[0][0].body).toMatchSnapshot()

  // Adds a message telling the user the comment was updated
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchInlineSnapshot(
    `"@somebody I found some new diffs since the last time I checked. Take a look at the comment to see what changed."`
  )
})

it('should notify the user when resolving a comment', async () => {
  listCommentsMock.mockReturnValue([{ body: sampleComment, node_id: 3 }])
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar' } },
    { 'en-US/translation': { foo: 'bar' } }
  )

  await runAction()

  // Minimizes the comment
  expect(graphqlMock).toHaveBeenCalledWith(expect.anything(), { subjectId: 3 })

  // Adds a message telling the user the comment was minimized
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchInlineSnapshot(
    `"@somebody Looks like there are no longer any diffs, so I went ahead and resolved the outdated comment."`
  )
})

it('should notify the user when there are no diffs', async () => {
  mockListResources((version) => [`projectId/${version}/en-US/translation`])
  mockFetchResource(
    { 'en-US/translation': { foo: 'bar' } },
    { 'en-US/translation': { foo: 'bar' } }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchInlineSnapshot(
    `"@somebody Good news! I didn't find any diffs so everything in Locize is up to date!"`
  )
})
