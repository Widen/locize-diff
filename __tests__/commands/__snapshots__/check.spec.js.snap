// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow manually requesting a diff 1`] = `
"Heads up! Looks like there are some differences between your two Locize versions.

### \`en-US/translation\`

| Key | \`latest\` | \`production\` |
| --- | --- | --- |
| \`foo\` | bar | baz |

_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._"
`;

exports[`should notify the user when updating an existing comment 1`] = `
"Heads up! Looks like there are some differences between your two Locize versions.

### \`en-US/translation\`

| Key | \`latest\` | \`production\` |
| --- | --- | --- |
| \`foo\` | bar | baz |
| \`hi\` | ho | howdy |

_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._"
`;
